lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp

```diff
@@ -180,12 +273,7 @@ class HoistLayoutConversion : public OpRewritePattern<ConvertLayoutOp> {
       if (isa<LoadOp>(currOp)) {
         foundLoad = true;
       } else if (foundLoad) {
-        // Bail out if there exists an op after Load that is not FpToFp,
-        // Bitcast, or Arith.
-        if (!isa<FpToFpOp, BitcastOp>(currOp) &&
-            !isPureUnaryInlineAsm(currOp) &&
-            currOp->getDialect()->getTypeID() !=
-                TypeID::get<arith::ArithDialect>())
+        if (!canHoistDotOpEncV2(currOp, dotOpEnc))
```

Comment:
The restriction here was more relaxed than the one above. I think we should only fail on the old conditions here.

Reply:
I thought that this was a latent bug, i.e. the below conditions were too relaxed, and the same strict conditions should be applied to all ops in the hoist-path, not just the last op before `local_alloc`.

For example, this check
```cpp
    // Don't hoist through u1 -> fp casts as they aren't supported in
    // ElementwiseOpToLLVM::reorderValues().
    if (isa<arith::UIToFPOp>(src)) {
      Type srcType = getElementTypeOrSelf(src->getOperand(0));
      if (srcType.isInteger(1))
        return failure();
    }
```
should be necessary for all ops everywhere, since we simply cannot lower tensors in dotOp layout with width == 1 to LLVM.

Same thing can probably be said of `SelectOp` and the rest (except for maybe `TruncF` and `TruncI`, which were removed/relaxed in another PR), or am I mistaken?

Reply:
I reviewed the code again. What you are saying makes sense yes. The restriction should be the same. I guess I commented on it just from the perspective of "behavior is changing" without deeper inspection. Thanks for your explanation.